overlord/devicestate/devicestate.go

```diff
@@ -393,39 +443,100 @@ func remodelTasks(ctx context.Context, st *state.State, current, new *asserts.Mo
+				// this is unexpected, new model has the same base,
+				// which should be already installed
+				return nil, err
+			}
+			if bs.TrackingChannel != newBaseChannel {
+				newBaseChannel = new.BaseSnap().DefaultChannel
+			}
 		}
-		tss = append(tss, ts)
-	}
-	if current.Gadget() != new.Gadget() {
-		ts, err := snapstateInstallWithDeviceContext(ctx, st, new.Gadget(), &snapstate.RevisionOptions{Channel: new.GadgetTrack()}, userID, snapstate.Flags{}, deviceCtx, fromChange)
-		if err != nil {
-			return nil, err
+		if newBaseChannel != "" {
+			ts, err := snapstateUpdateWithDeviceContext(st, new.Base(),
+				&snapstate.RevisionOptions{Channel: newBaseChannel},
+				userID, snapstate.Flags{NoReRefresh: true}, deviceCtx, fromChange)
+			if err != nil {
+				return nil, err
+			}
+			tss = append(tss, ts)
 		}
-		tss = append(tss, ts)
 	}
+	// gadget
+	gadgetChanged, newGadgetChannel := gadgetChange(current, new)
+	if gadgetChanged {
+		if current.Gadget() == new.Gadget() {
+			ts, err := snapstateUpdateWithDeviceContext(st, new.Gadget(),
+				&snapstate.RevisionOptions{Channel: newGadgetChannel},
+				userID, snapstate.Flags{NoReRefresh: true}, deviceCtx, fromChange)
+			if err != nil {
+				return nil, err
+			}
+			tss = append(tss, ts)
+		} else {
+			ts, err := snapstateInstallWithDeviceContext(ctx, st, new.Gadget(),
+				&snapstate.RevisionOptions{Channel: newGadgetChannel},
+				userID, snapstate.Flags{}, deviceCtx, fromChange)
+			if err != nil {
+				return nil, err
+			}
+			tss = append(tss, ts)
+		}
 
-	// add new required-snaps, no longer required snaps will be cleaned
-	// in "set-model"
-	for _, snapRef := range new.RequiredNoEssentialSnaps() {
+	}
+	// go through all the model snaps, see if there are new required snaps
+	// or a track for existing ones needs to be updated
+	for _, modelSnap := range new.SnapsWithoutEssential() {
 		// TODO|XXX: have methods that take refs directly
 		// to respect the snap ids
-		needsInstall, err := notInstalled(st, snapRef.SnapName())
+		currentInfo, err := snapstate.CurrentInfo(st, modelSnap.SnapName())
+		needsInstall := false
 		if err != nil {
-			return nil, err
+			if !isNotInstalled(err) {
+				return nil, err
+			}
+			if modelSnap.Presence == "required" {
+				needsInstall = true
+			}
 		}
+		// default channel can be set only in UC20 models
 		if needsInstall {
 			// If the snap is not installed we need to install it now.
-			ts, err := snapstateInstallWithDeviceContext(ctx, st, snapRef.SnapName(), nil, userID, snapstate.Flags{Required: true}, deviceCtx, fromChange)
+			ts, err := snapstateInstallWithDeviceContext(ctx, st, modelSnap.SnapName(),
+				&snapstate.RevisionOptions{Channel: modelSnap.DefaultChannel},
+				userID,
+				snapstate.Flags{Required: true}, deviceCtx, fromChange)
 			if err != nil {
 				return nil, err
 			}
 			tss = append(tss, ts)
+		} else if currentInfo != nil && modelSnap.DefaultChannel != "" {
+			var ss snapstate.SnapState
```

Comment:
this code path doesn't seem to be tested atm

Reply:
this is tested in the managers suite, testRemodelUC20WithRecoverySystem